src/effects/native/echoeffect.cpp

```diff
@@ -95,42 +110,63 @@ void EchoEffect::processChannel(const ChannelHandle& handle, EchoGroupState* pGr
                                 const GroupFeatureState& groupFeatures) {
     Q_UNUSED(handle);
-    Q_UNUSED(groupFeatures);
+
     DEBUG_ASSERT(0 == (numSamples % EchoGroupState::kChannelCount));
     EchoGroupState& gs = *pGroupState;
-    double delay_time = m_pDelayParameter->value();
+    // The minimum of the parameter is zero so the exact center of the knob is 1 beat.
+    double period = m_pDelayParameter->value();
     double send_amount = m_pSendParameter->value();
     double feedback_amount = m_pFeedbackParameter->value();
     double pingpong_frac = m_pPingPongParameter->value();
 
-    int delay_samples = EchoGroupState::kChannelCount * delay_time * sampleRate;
+    int delay_samples;
+    if (m_pSyncParameter->toBool() && groupFeatures.has_beat_length) {
+        // period is a number of beats
+        if (period < 0.125) {
+            period = 0.125;
+        } else {
+            period = roundToFraction(period, 4);
+        }
+        delay_samples = period * groupFeatures.beat_length;
+    } else {
+        // period is a number of seconds
+        period = std::max(period, 0.125);
+        delay_samples = period * sampleRate * EchoGroupState::kChannelCount;
+    }
+    VERIFY_OR_DEBUG_ASSERT(delay_samples > 0) {
+        delay_samples = 1;
+    }
     VERIFY_OR_DEBUG_ASSERT(delay_samples <= gs.delay_buf.size()) {
         delay_samples = gs.delay_buf.size();
     }
 
-    if (delay_time < gs.prev_delay_time) {
+    if (period < gs.prev_period) {
         // If the delay time has shrunk, we may need to wrap the write position.
         gs.write_position = gs.write_position % delay_samples;
-    } else if (delay_time > gs.prev_delay_time) {
+    } else if (period > gs.prev_period) {
         // If the delay time has grown, we need to zero out the new portion
         // of the buffer we are using.
         SampleUtil::applyGain(gs.delay_buf.data(gs.prev_delay_samples), 0,
                               gs.delay_buf.size() - gs.prev_delay_samples);
     }
 
     int read_position = gs.write_position;
-    gs.prev_delay_time = delay_time;
+    gs.prev_period = period;
     gs.prev_delay_samples = delay_samples;
 
     // Feedback the delay buffer and then add the new input.
     for (unsigned int i = 0; i < numSamples; i += EchoGroupState::kChannelCount) {
         // Ramp the beginning and end of the delay buffer to prevent clicks.
         double write_ramper = 1.0;
-        if (gs.write_position < EchoGroupState::kRampLength) {
-            write_ramper = static_cast<double>(gs.write_position) / EchoGroupState::kRampLength;
-        } else if (gs.write_position > delay_samples - EchoGroupState::kRampLength) {
-            write_ramper = static_cast<double>(delay_samples - gs.write_position)
-                    / EchoGroupState::kRampLength;
+        // When feedback is 1.0, continously applying ramping to the buffer creates
+        // an unpleasant artifact.
+        if (feedback_amount != 1.0) {
```

Comment:
It looks like the ramping code is wrong here. From the theory, an echo is a steady stream, echoing the original stream with a delay. This means that we do not apply a ramp if the effect is settled. What we need is a ramp that is a applied to the first samples in the buffer. I think that can be done by ramping the send_amount from zero after starting the effect. Changes in send_amount and feedback_amount is required anyway, to avoid clicks when changing.